ftp-source/src/main/java/org/springframework/cloud/stream/module/ftp/FtpSource.java

```diff
@@ -77,15 +82,24 @@ public PollerMetadata poller() {
 
 	@Bean
 	public IntegrationFlow ftpInboundFlow() {
-		IntegrationFlowBuilder flowBuilder = IntegrationFlows.from(Ftp.inboundAdapter(FtpSource.this.ftpSessionFactory)
+		FtpInboundChannelAdapterSpec messageSourceBuilder = Ftp.inboundAdapter(FtpSource.this.ftpSessionFactory)
 				.preserveTimestamp(config.isPreserveTimestamp())
 				.remoteDirectory(config.getRemoteDir())
-				.patternFilter(config.getFilenamePattern())
 				.remoteFileSeparator(config.getRemoteFileSeparator())
 				.localDirectory(config.getLocalDir())
 				.autoCreateLocalDirectory(config.isAutoCreateLocalDir())
 				.temporaryFileSuffix(config.getTmpFileSuffix())
-				.deleteRemoteFiles(config.isDeleteRemoteFiles())
+				.deleteRemoteFiles(config.isDeleteRemoteFiles());
+
+		if (StringUtils.hasText(this.config.getFilenamePattern())) {
+			messageSourceBuilder.filter(new FtpSimplePatternFileListFilter(this.config.getFilenamePattern()));
+		}
+		else if (StringUtils.hasText(this.config.getFilenameRegex())) {
+			messageSourceBuilder
+					.filter(new FtpRegexPatternFileListFilter(Pattern.compile(this.config.getFilenameRegex())));
```

Comment:
Should make the property a `Pattern` itself (adding a dedicated Converter if none already exists). This would surface pattern compilation errors early